README
NAME
    "Socket::Netlink" - interface to Linux's "PF_NETLINK" socket family

SYNOPSIS
     use Socket;
     use Socket::Netlink qw( :DEFAULT pack_nlmsghdr unpack_nlmsghdr );

     socket( my $sock, PF_NETLINK, SOCK_RAW, 0 ) or die "socket: $!";

     send( $sock, pack_nlmsghdr( 18, NLM_F_REQUEST|NLM_F_DUMP, 0, 0,
          "\0\0\0\0\0\0\0\0" ), 0 )
        or die "send: $!";

     recv( $sock, my $buffer, 65536, 0 ) or die "recv: $!";

     printf "Received type=%d flags=%x:\n%v02x\n",
        ( unpack_nlmsghdr( $buffer ) )[ 0, 1, 4 ];

DESCRIPTION
    This module contains the low-level constants and structure handling
    functions required to use Linux's "PF_NETLINK" socket family. It is
    suggested to use the high-level object interface to this instead; see
    IO::Socket::Netlink.

CONSTANTS
    The following constants are exported

    PF_NETLINK
            The packet family (for "socket()" calls)

    AF_NETLINK
            The address family

ADDRESS FUNCTIONS
    The following pair of functions operate on "AF_NETLINK" address
    structures. The meainings of the parameters are:

    pid     The unique endpoint number for this netlink socket. If given as
            0 to the "bind()" syscall, the kernel will allocate an endpoint
            number of the process's PID.

    groups  A 32-bit bitmask of the multicast groups to join.

  $addr = pack_sockaddr_nl( $pid, $groups )
    Returns a "sockaddr_nl" structure with the fields packed into it.

  ( $pid, $groups ) = unpack_sockaddr_nl( $addr )
    Takes a "sockaddr_nl" structure and returns the unpacked fields from it.

STRUCTURE FUNCTIONS
    The following function pairs operate on structure types used by netlink

  $buffer = pack_nlmsghdr( $type, $flags, $seq, $pid, $body )
  ( $type, $flags, $seq, $pid, $body, $morebuffer ) = unpack_nlmsghdr( $buffer )
    Pack or unpack a "struct nlmsghdr" and its payload body.

    Because a single netlink message can contain more than payload body, the
    "unpack_nlmsghdr" function will return the remaining buffer after
    unpacking the first message, in case there are others. If there are no
    more, the $morebuffer list element will not be returned.

     while( defined $buffer ) {
        ( my ( $type, $flags, $seq, $pid, $body ), $buffer ) = unpack_nlmsghdr( $buffer );
        ...
     }

    There is no similar functionallity for "pack_nlmsghdr"; simply
    concatenate multiple results together to send more than one message.

  $buffer = pack_nlmsgerr( $error, $msg )
  ( $error, $msg ) = unpack_nlmsgerr( $buffer )
    Pack or unpack a "struct nlmsgerr". The kernel expects or reports
    negative integers in its structures; these functions take or return
    normal positive error values suitable for use with $!.

  $buffer = pack_nlattrs( %attrs )
  %attrs = unpack_nlattrs( $buffer )
    Pack or unpack a list of netlink attributes.

    These functions take or return even-sized lists of "$type, $value"
    pairs. The type will be the number in the netlink attribute message, and
    the value will be a plain packed string buffer. It is the caller's
    responsibilty to further pack/unpack this buffer as appropriate for the
    specific type.

    Because these functions take/return even-sized lists, they may be passed
    or returned into hashes.

SEE ALSO
    *   netlink(7) - netlink - Communication between kernel and userspace
        (AF_NETLINK)

    *   IO::Socket::Netlink - Object interface to "AF_NETLINK" domain
        sockets

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Socket-Netlink

0.04    CHANGES:
         * Optionally compile against some genetlink.h constants only found
           post-2.6.18 so as not to fail to build on older systems
         * Only upbless Netlink::Generic messages if they're not already a
           more specific subclass
         * Reuseable support for nlmsg_type dependent subclassing of message
           objects
         * Support different message classes for user->kernel commands vs.
           kernel->user notifications
         * Don't hard-depend on Sub::Uplevel

0.03    CHANGES:
         * Added accessors for finer access of individual netlink attribute
           values in messages

0.02    CHANGES:
         * Improved NLMSG_ERROR handling semantics
         * Don't cache NETLINK_GENERIC familyname => id mapping

        BUGFIXES:
         * Declare ExtUtils::CChecker as a configure_requires

0.01    First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;

use Module::Build;

eval { require ExtUtils::CChecker; 1 } or
   die "OS unsupported - missing ExtUtils::CChecker";

my $cc = ExtUtils::CChecker->new;

$cc->assert_compile_run(
   diag => "no PF_NETLINK",
   source => <<'EOF' );
#include <stdio.h>
#include <sys/socket.h>
int main(int argc, char *argv[]) {
  printf("PF_NETLINK is %d\n", PF_NETLINK);
  return 0;
}
EOF

# Some userland headers are sufficiently old that they don't have NLA_HDRLEN
# or struct nlattr. If this is the case we have little choice but to cheat a
# little and try to use the kernel ones instead.
chomp( my $uname_r = `uname -r` );

$cc->find_include_dirs_for(
   diag => "no NLA_HDRLEN",
   dirs => [ [], [ "/lib/modules/$uname_r/build/include" ] ],
   source => <<'EOF' );
#include <sys/socket.h>
#include <linux/netlink.h>
struct nlattr nla;
int len = NLA_HDRLEN;
int main(int argc, char *argv[]) {
  return 0;
}
EOF

my $build = $cc->new_module_build(
   module_name => 'Socket::Netlink',
   requires => {
      'Exporter' => '5.57',
   },
   configure_requires => {
      'ExtUtils::CChecker' => '0.02',
      'Module::Build' => 0,
      'Module::Build::Compat' => 0,
   },
   build_requires => {
      'ExtUtils::H2PM' => '0.07',
      'Module::Build' => 0,
      'Module::Build::Compat' => 0,
      'Test::HexString' => 0,
      'Test::More' => 0,
   },
   license => 'perl',
   create_makefile_pl => 'small',
   create_license => 1,
   create_readme  => 1,
);

$build->create_build_script;


lib/Socket/Netlink/Generic.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2010-2011 -- leonerd@leonerd.org.uk

package Socket::Netlink::Generic;

use strict;
use warnings;

our $VERSION = '0.04';

use Exporter 'import';

use Socket::Netlink::Generic_const;

=head1 NAME

C<Socket::Netlink::Generic> - interface to Linux's C<NETLINK_GENERIC> netlink
socket protocol

=head1 SYNOPSIS

 use Socket;
 use Socket::Netlink qw( :DEFAULT
    pack_nlmsghdr unpack_nlmsghdr pack_nlattrs unpack_nlattrs );
 use Socket::Netlink::Generic qw( :DEFAULT pack_genlmsghdr unpack_genlmsghdr );

 socket( my $sock, PF_NETLINK, SOCK_RAW, NETLINK_GENERIC ) or die "socket: $!";

 send( $sock, pack_nlmsghdr( NETLINK_GENERIC, NLM_F_REQUEST, 0, 0,
                 pack_genlmsghdr( CTRL_CMD_GETFAMILY, 0,
                    pack_nlattrs( CTRL_ATTR_FAMILY_NAME, "TASKSTATS\0" )
                 ),
              ),
    0 ) or die "send: $!";

 recv( $sock, my $buffer, 65536, 0 ) or die "recv: $!";

 my %attrs = unpack_nlattrs( 
               (unpack_genlmsghdr( (unpack_nlmsghdr $buffer )[4] ) )[2]
             );

 printf "TASKSTATS family ID is %d\n",
    unpack( "S", $attrs{CTRL_ATTR_FAMILY_ID()} );

=head1 DESCRIPTION

This module contains the low-level constants and structure handling functions
required to use the C<NETLINK_GENERIC> protocol of Linux's C<PF_NETLINK>
socket family. It is suggested to use the high-level object interface to this
instead; see L<IO::Socket::Netlink::Generic>.

=cut

=head1 CONSTANTS

The following sets of constants are exported:

The netlink protocol constant:

 NETLINK_GENERIC

Control commands:

 CTRL_CMD_NEWFAMILY    CTRL_CMD_DELFAMILY    CTRL_CMD_GETFAMILY
 CTRL_CMD_NEWOPS       CTRL_CMD_DELOPS       CTRL_CMD_GETOPS
 CTRL_CMD_NEWMCAST_GRP CTRL_CMD_DELMCAST_GRP CTRL_CMD_GETMCAST_GRP

Attribute IDs:

 CTRL_ATTR_FAMILY_ID    CTRL_ATTR_FAMILY_NAME CTRL_ATTR_VERSION
 CTRL_ATTR_HDRSIZE      CTRL_ATTR_MAXATTR     CTRL_ATTR_OPS
 CTRL_ATTR_MCAST_GROUPS

Nested attribute IDs:

 CTRL_ATTR_OP_ID CTRL_ATTR_OP_FLAGS
 CTRL_ATTR_MCAST_GRP_NAME CTRL_ATTR_MCAST_GRP_ID

Note that if the kernel headers are particularly old, not all of these
constants may be available. If they are unavailable at compile time, no
constant functions will be generated.

=cut

=head1 STRUCTURE FUNCTIONS

=head2 $buffer = pack_genlmsghdr( $cmd, $version, $body )

=head2 ( $cmd, $version, $body ) = unpack_genlmsghdr( $buffer )

Pack or unpack a C<struct genlmsghdr> and its payload body.

=cut

=head1 SEE ALSO

=over 4

=item *

L<Socket::Netlink> - interface to Linux's C<PF_NETLINK> socket family

=item *

L<IO::Socket::Netlink::Generic> - Object interface to C<NETLINK_GENERIC>
netlink protocol sockets

=back

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Socket/Netlink.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2010 -- leonerd@leonerd.org.uk

package Socket::Netlink;

use strict;
use warnings;

use Carp;

our $VERSION = '0.04';

use Exporter 'import';
our @EXPORT_OK = qw(
   pack_sockaddr_nl unpack_sockaddr_nl
   pack_nlmsghdr    unpack_nlmsghdr
   pack_nlmsgerr    unpack_nlmsgerr
   pack_nlattrs     unpack_nlattrs
);

require XSLoader;
XSLoader::load( __PACKAGE__, $VERSION );

=head1 NAME

C<Socket::Netlink> - interface to Linux's C<PF_NETLINK> socket family

=head1 SYNOPSIS

 use Socket;
 use Socket::Netlink qw( :DEFAULT pack_nlmsghdr unpack_nlmsghdr );

 socket( my $sock, PF_NETLINK, SOCK_RAW, 0 ) or die "socket: $!";

 send( $sock, pack_nlmsghdr( 18, NLM_F_REQUEST|NLM_F_DUMP, 0, 0,
      "\0\0\0\0\0\0\0\0" ), 0 )
    or die "send: $!";

 recv( $sock, my $buffer, 65536, 0 ) or die "recv: $!";

 printf "Received type=%d flags=%x:\n%v02x\n",
    ( unpack_nlmsghdr( $buffer ) )[ 0, 1, 4 ];

=head1 DESCRIPTION

This module contains the low-level constants and structure handling functions
required to use Linux's C<PF_NETLINK> socket family. It is suggested to use
the high-level object interface to this instead; see L<IO::Socket::Netlink>.

=cut

=head1 CONSTANTS

The following constants are exported

=over 8

=item PF_NETLINK

The packet family (for C<socket()> calls)

=item AF_NETLINK

The address family

=back

=cut

=head1 ADDRESS FUNCTIONS

The following pair of functions operate on C<AF_NETLINK> address structures.
The meainings of the parameters are:

=over 8

=item pid

The unique endpoint number for this netlink socket. If given as 0 to the
C<bind()> syscall, the kernel will allocate an endpoint number of the
process's PID.

=item groups

A 32-bit bitmask of the multicast groups to join.

=back

=head2 $addr = pack_sockaddr_nl( $pid, $groups )

Returns a C<sockaddr_nl> structure with the fields packed into it.

=head2 ( $pid, $groups ) = unpack_sockaddr_nl( $addr )

Takes a C<sockaddr_nl> structure and returns the unpacked fields from it.

=cut

=head1 STRUCTURE FUNCTIONS

The following function pairs operate on structure types used by netlink

=head2 $buffer = pack_nlmsghdr( $type, $flags, $seq, $pid, $body )

=head2 ( $type, $flags, $seq, $pid, $body, $morebuffer ) = unpack_nlmsghdr( $buffer )

Pack or unpack a C<struct nlmsghdr> and its payload body.

Because a single netlink message can contain more than payload body, the
C<unpack_nlmsghdr> function will return the remaining buffer after unpacking
the first message, in case there are others. If there are no more, the
C<$morebuffer> list element will not be returned.

 while( defined $buffer ) {
    ( my ( $type, $flags, $seq, $pid, $body ), $buffer ) = unpack_nlmsghdr( $buffer );
    ...
 }

There is no similar functionallity for C<pack_nlmsghdr>; simply concatenate
multiple results together to send more than one message.

=head2 $buffer = pack_nlmsgerr( $error, $msg )

=head2 ( $error, $msg ) = unpack_nlmsgerr( $buffer )

Pack or unpack a C<struct nlmsgerr>. The kernel expects or reports negative
integers in its structures; these functions take or return normal positive
error values suitable for use with C<$!>.

=head2 $buffer = pack_nlattrs( %attrs )

=head2 %attrs = unpack_nlattrs( $buffer )

Pack or unpack a list of netlink attributes.

These functions take or return even-sized lists of C<$type, $value> pairs.
The type will be the number in the netlink attribute message, and the value
will be a plain packed string buffer. It is the caller's responsibilty to
further pack/unpack this buffer as appropriate for the specific type.

Because these functions take/return even-sized lists, they may be passed or
returned into hashes.

=cut

=head1 SEE ALSO

=over 4

=item *

C<netlink(7)> - netlink - Communication between kernel and userspace (AF_NETLINK)

=item *

L<IO::Socket::Netlink> - Object interface to C<AF_NETLINK> domain sockets

=back

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
